packages/bpk-scrim-utils/src/withScrimmedPortal.tsx

```diff
@@ -37,22 +42,34 @@ const getPortalElement = (target: (() => HTMLElement | null) | null | undefined)
     if (document.body) {
         return document.body;
     }
-    throw new Error('Render target and fallback unavailable');
+    throw new Error('Render target and fallback unavailable. The ScrimmedComponent uses a portal making it unsuitable for SSR. If you are trying to render on SSR, please pass the `runOnServer` prop.');
 }
 
 const withScrimmedPortal = (WrappedComponent: ComponentType<ScrimProps>) => {
     const Scrimmed = withScrim(WrappedComponent);
 
-    const ScrimmedComponent = ({ renderTarget, ...rest}: Props) => {
-        const portalElement = getPortalElement(renderTarget);
-
+    const ScrimmedComponent = ({ renderTarget, runOnServer = false, ...rest}: Props) => {
         const [isPortalReady, setIsPortalReady] = useState(false);
 
         useEffect(() => {
             setIsPortalReady(true);
           }, []);
 
-        return createPortal(<Scrimmed {...rest} isPortalReady={isPortalReady} />, portalElement);
+        /**
+         * The following code runs only on the client.
+         * If consumer passes runOnServer as false (meaning they want to render the component on client only), we render the component as normal.
+         * If consumer passes runOnServer as true (meaning they want to render the component on server and client), we render the component as normal once the component has been mounted.
+         * This is to ensure the snapshotted markup (initial render before the component has been mounted) is the same on both server and client.
+         */
+        if (!runOnServer || isPortalReady) {
```

Comment:
I'm a bit confused about this, what really decides if we enter here is the `isPortalReady`, what's the purpose of `runOnServer` exactly?

Reply:
> what's the purpose of runOnServer exactly?

Because if you want to render this component only on client side, then you'll just run this code directly.
However, if you want to render it on both client and server, the behaviour will be slightly different on client - on the first render it will be rendering the `BpkScrim`, and then once it's been mounted it will be rendering the portal.
That is because if you want to render on both server and client, the initial render needs to be exactly the same, otherwise we get a warning from React and might run into some unexpected issues. But if we only render on client, we don't need to care about it so we just render it normally

Reply:
Thanks for the comment @anambl, I think I understand the challenge you want to solve. This double render feels sub-optimal, wondering if this is a recommended approach from somewhere?

What's the specific warning you get? Is it a prod warning or just local dev?

Reply:
I think it's actually okay as the `BpkScrim` is rendered as part of the `withScrim` anyways so all we do is add some new things once the component has been mounted 🤔
Most issues I've found will recommend in these cases to basically only return something in the component if it has been mounted 😅 so something like this https://github.com/vercel/next.js/discussions/17443#discussioncomment-637879

The warning is `Warning: Expected server HTML to contain a matching <div> in <div>`

Reply:
> I think it's actually okay

I don't doubt that it's _okay_ :) my concern is the double render is emitting a big code smell to me and I can imagine this run around will be hard to understand for maintainers in months and years to come

One question on this comment:

>  if you want to render it on both client and server, the behaviour will be slightly different on client - on the first render it will be rendering the BpkScrim

What's the difference between what's expected to be rendered on the server and in the client?

Reply:
Did you look at just [suppressing the warning](https://react.dev/reference/react-dom/hydrate#suppressing-unavoidable-hydration-mismatch-errors)? 

It's an expected difference, so as long as we add code comments explaining the suppression I think we've covered the requirement here?

Reply:
> I'm a little bit confused about what you mean by double render

<img width="933" alt="image" src="https://github.com/Skyscanner/backpack/assets/1532576/cf987f18-1757-48a0-a4fa-04e8c48e46a6">

We're doing this, but in a weird round about way to get around a react warning that we can supress

>  I don't think this counts as an expected difference 

By expected, I mean the component is set up to render something different on the server and client. It doesn't matter what that difference is, the DOM is different, and react knows it's different because it's throwing the warning.

We're going round in circles here, and I think it's clear me and @arnaugm have reservations about this aproach.

To try and get a resolution quickly, is there a reason you don't want to or can't suppress the warning?

Reply:
> We're doing this, but in a weird round about way to get around a react warning that we can supress

Why do you think it's in a weird round about way? I think it is exactly what the example is doing. The only difference is that we have this extra prop which consumers can pass if they render their component **only** on the client, but it's not that important and I can remove it if it causes confusion.

> By expected, I mean the component is set up to render something different on the server and client. 

In that sense it is expected, but the React team advises suppressing the warning as the last resort and their example actually makes sense because they're using timestamps which of course will be different. Our use case is different imo. The recommended approach is to render the same content on the initial render on server and client (which I had linked [above](https://react.dev/reference/react-dom/hydrate#handling-different-client-and-server-content)). There's a few articles about this, this one specifically compares these exact 2 solutions: https://blog.hao.dev/render-client-side-only-component-in-next-js

> To try and get a resolution quickly, is there a reason you don't want to or can't suppress the warning?

I am definitely in favour of _not_ suppressing any warnings, but I do agree we need to get a resolution quickly so if the preferred solution from both of you is to suppress it and no one else has strong feelings about it then it's fine.

Reply:
What does this PR look like if we remove the `runOnServer` prop but keep the `useEffect` on mount? 

Does react show the warning?

Reply:
> Does react show the warning?

No, it does not.

So the purpose of the `runOnServer` prop (which I agree it needs to have a different name) was simply to make the default of this component be client-only as I think that makes most sense given it's using a portal. As you said earlier, banana is CSR only, but there's also projects that do conditional rendering and render some components CSR only so by having this prop by default the component will be working on client only.
But, if you want to run it on both client and server, you can pass this prop so on server you'll be getting the scrim first. The prop has nothing to do with the React warning. I've added it because I wanted this component to be client-only by default, but if it's rendered in a project that uses SSR, you can use it there too and you'll be getting the scrim on the server. So if we remove the prop, on CSR-only projects it will still be rendering the scrim first and then the component which doesn't really make much sense given we only need to do it if we hydrate, but that's not a big issue as it wouldn't even be visible to the user so I'm ok with removing it if it causes confusion

Reply:
> The prop has nothing to do with the React warning

Thanks for this clarification, I was under the impression the prop and react warning were linked.

> if we remove the prop, on CSR-only projects it will still be rendering the scrim first and then the component which doesn't really make much sense given we only need to do it if we hydrate, but that's not a big issue as it wouldn't even be visible to the user so I'm ok with removing it if it causes confusion

I understand your concern here, but I think it's a case of premature optimisation. As you say users won't notice this, but by adding this prop for CSR-only is introducing a lot of confusing logic (as evidenced by the comments on this PR). If we're comparing these two concerns, my preference is the one with the better DX.

Reply:
I think like Mungo suggested above, the prop can just be replaced with checking if `window` is defined so we can achieve the same and remove the prop